pkg/tenant/apikeys.go

```diff
@@ -1,36 +1,169 @@
-	// implemented for each endpoint.
+	// User credentials are required to make the Quarterdeck request
+	if ctx, err = middleware.ContextFromRequest(c); err != nil {
+		log.Error().Err(err).Msg("could not create user context from request")
+		c.JSON(http.StatusUnauthorized, api.ErrorResponse("could not fetch credentials for authenticated user"))
+		return
+	}
 
-	// TODO: Add authentication and authorization middleware
-	// TODO: Identify top-level info
-	// TODO: Parse and validate user input
-	// TODO: Perform work on the request, e.g. database interactions,
-	// sending notifications, accessing other services, etc.
+	// Parse the params from the POST request
+	params := &api.APIKey{}
+	if err = c.BindJSON(params); err != nil {
+		log.Warn().Err(err).Msg("could not parse API key params")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("could not parse API key params"))
+		return
+	}
 
-	// Return response with the correct status code
+	// Name is required
+	if params.Name == "" {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("API key name is required"))
+		return
+	}
 
-	// TODO: Replace StatusNotImplemented with StatusOk and
-	// replace "not yet implemented" message.
+	// Permissions are required
+	if len(params.Permissions) == 0 {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("API key permissions are required"))
+		return
+	}
+
+	// Build the Quarterdeck request
+	// See ValidateCreate() for required fields
+	req := &qd.APIKey{
+		Name:        params.Name,
+		Permissions: params.Permissions,
+	}
+
+	// ProjectID is required
+	projectID := c.Param("projectID")
+	if req.ProjectID, err = ulid.Parse(projectID); err != nil {
+		log.Warn().Err(err).Str("projectID", projectID).Msg("could not parse project ID")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("invalid project ID"))
+		return
+	}
+
+	// TODO: Validate that the user is associated with the project by checking the
+	// orgID in the claims against the project's orgID
+
+	// TODO: Add source to request
+
+	// Create the API key with Quarterdeck
+	var key *qd.APIKey
+	if key, err = s.quarterdeck.APIKeyCreate(ctx, req); err != nil {
+		log.Error().Err(err).Msg("could not create API key")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not create API key"))
+		return
+	}
+
+	// Return the API key
+	out := &api.APIKey{
+		ID:           key.ID.String(),
+		ClientID:     key.ClientID,
+		ClientSecret: key.ClientSecret,
+		Name:         key.Name,
+		Owner:        key.CreatedBy.String(),
+		Permissions:  key.Permissions,
+		Created:      key.Created.Format(time.RFC3339Nano),
+		Modified:     key.Modified.Format(time.RFC3339Nano),
+	}
+
+	c.JSON(http.StatusCreated, out)
+}
+
+// TODO: Implement by factoring out common code from ProjectAPIKeyCreate
+func (s *Server) APIKeyList(c *gin.Context) {
 	c.JSON(http.StatusNotImplemented, "not implemented yet")
 }
 
+// TODO: Implement by factoring out common code from ProjectAPIKeyCreate
```

Comment:
While reviewing older correspondence, I thought that API Keys would only be created within a project. This is why the client APIKeyCreate was previously removed along with the client info removed (sorry!).

Reply:
Right, I think there was (and might still be) some uncertainty about that. I think the original idea was that API keys could be created on either the organization level or within a specific project. It doesn't really matter to quarterdeck however since the projectID is an optional parameter in quarterdeck's CreateAPIKey handler.